ACPs/99-validatorsetmanager-contract/README.md

```diff
@@ -1,263 +1,343 @@
 ```
 
-- The Subnet ID
-- The NodeID
-- The `MessageID` corresponding to the Warp message in the `RegisterSubnetValidatorTx` that added the validator (SHA256)
-- The `nonce` corresponding to this update
-- The weight of the validator (if `0`, this means that the validation is over)
-- The exact timestamp at which the weight was updated (if the nonce is `0`, this also means the beginning of the `validation`)
-- The uptime observed during the previous validation period (only useful in case a validator weight update)
-
-Note that the timestamp can also be used to update the `endTime` of the previous validation period.
-
-**Note:** Once this discussion/ACP is in an advanced stage, an update to ACP-77 can be made to specify the payload of P-Chain notifications.
-
-### Reference Architecture for Custom “Security Modules”
-
-The `ValidatorSetManager` contract will be ownable (e.g. using [OpenZeppelin Ownable](https://docs.openzeppelin.com/contracts/5.x/api/access#Ownable)). The ownership can be transferred to a smart contract, the “Security Module”, that will implement the security rules for the Subnet.
-
-![ValidatorSetManager architecture](./validatorsetmanager-architecture.png)
-
-The most simple architecture to implement a PoA Subnet would be to use a [Safe Multisig (or Smart Account)](https://docs.safe.global/advanced/smart-account-overview) as owner, effectively reproducing the existing Subnet management functionalities available on the P-Chain today.
-
-Other “Security Modules” could implement PoS, Liquid PoS, etc. The specification of such smart contracts is out of the scope of this ACP.
+"Security modules" could implement PoS, Liquid PoS, etc. The specification of such smart contracts is out of the scope of this ACP.
 
 ## Backwards Compatibility
 
-The `ValidatorSetManager` is only a reference implementation, it doesn’t have any impact on the current behavior of the protocol.
+The `IACP99Manager` and `IACP99SecurityModule` interfaces are only a reference interface, they don’t have any impact on the current behavior of the Avalanche protocol.
 
 ## Reference Implementation
 
-A full reference implementation has not been provided yet. It will be provided once this ACP is considered `Implementable`.
+A work in progress reference implementation is available in the [Suzaku Contracts Library](https://github.com/suzaku-network/suzaku-contracts-library/blob/acp-99-implementation/README.md#acp99-contracts-library) repository. It will be updated until this ACP is considered `Implementable` based on the outcome of the discussion.
 
 ## Security Considerations
 
 The audit process of the `ValidatorSetManager` contract is of the utmost importance for the future of the Avalanche ecosystem as most Subnets would rely upon it to secure their Subnet.
 
 ## Open Questions
 
-### Should the Subnet validators’ uptime be tracked on the P-Chain?
+### Is there an interest to keep historical information about the validator set on the manager chain?
```

Comment:
**Answer to this question (no need to answer here maybe I'll just post in the discussion when ACP is updated):**

On a PoS L1 with rewards depending on the weight and not only on uptime, when would the rewards be computed?  Maybe in this case it would practical to have the history somewhere in the `Manager` rather that having to index/parse the P-Chain?. That's the only use case I can think of.

Reply:
Computing the rewards is up to the security module. The security module can persist the [ValidatorWeightChangeInfo](https://github.com/suzaku-network/suzaku-contracts-library/blob/4cb200d6dfbf08430dab330ae0f1291be4b7440c/contracts/src/interfaces/ACP99/IACP99SecurityModule.sol#L44) if needed, for example at the end of every validation.